### PR TITLE
Exogw 342

### PR DESCRIPTION
**Pull request by steve-keep**
This is a major update that removes the dependency of Type GraphQL.

### CODE REVIEW

**Comment by steve-keep** [comment]
@coderabbitai review

File: src/packages/core/src/resolvers.ts [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+		| GraphQLEntityConstructor<G, D>[];
+	let isList = false;
+
+	if (Array.isArray(gqlEntityType)) {
+		isList = true;
+		gqlEntityType = gqlEntityType[0];
+	}
+
+	// @todo: Should the user specified filter be and-ed here?
+	//        My worry is if we just pass the filter through, it could be used to circumvent the relationship join.
```
Comment (steve-keep) [inline src/packages/core/src/resolvers.ts:393]: Should we start by anding? That way we know our filter is always applied, I think that is how I assumed it would work when using it as you can potentially get data that is not related to the parent.

Comment (thekevinbrown) [reply src/packages/core/src/resolvers.ts:393]: Yeah I'm down with that.

Comment (steve-keep) [reply src/packages/core/src/resolvers.ts]: done 🚀

File: src/packages/storage-provider/src/storageProvider.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		});
+	}
+
+	async getUploadUrl(
+		_source: Source,
+		{ key }: { key: string },
+		_ctx: BaseContext,
+		_info: GraphQLResolveInfo
+	): Promise<{ url: string; filename: string; type: MediaType }> {
+		if (!this.bucketName) throw new Error('Missing required env AWS_S3_BUCKET');
```
Comment (steve-keep) [inline src/packages/storage-provider/src/storageProvider.ts:82]: There are no hook calls or ACL checks on these mutations and query.

Should we add them as part of this work or create a new task to track it?

Comment (thekevinbrown) [reply src/packages/storage-provider/src/storageProvider.ts:82]: Let's go ahead and finish this properly as part of this work.

File: src/packages/aws-cognito/src/util.ts [hunk 3]
```diff
@@ -47,7 +47,7 @@ export const createUser = async (
 	if (!user) return null;
 	return {
 		...user,
-		Groups: groups.split(','),
+		Groups: groups?.split(','),
```
Comment (thekevinbrown) [inline src/packages/aws-cognito/src/util.ts:50]: Might be a bit nicer to do:
```suggestion
		Groups: groups?.split(',') ?? [],
```

File: src/packages/core/src/schema-builder.ts [hunk 4]
```diff
@@ -0,0 +1,832 @@
...
+		});
+
+		updateTypes.set(entity.name, updateType);
+	}
+
+	return updateType;
+};
+
+export interface SchemaBuilderOptions {
+	authChecker?: AuthChecker<any, any>;
```
Comment (thekevinbrown) [inline src/packages/core/src/schema-builder.ts:510]: I think this is a hangover from TypeGraphQL. Is this actually getting called anywhere now?

File: src/packages/core/src/types.ts [hunk 5]
```diff
@@ -199,3 +199,62 @@ export interface BackendProviderConfig {
...
+	complexity?: Complexity;
+	defaultValue?: any;
+
+	// This marks the field as read only in both the API and the admin UI.
+	// This will supersede any other read only settings.
+	readonly?: boolean;
+
+	nullable?: boolean | 'items' | 'itemsAndList';
+	excludeFromFilterType?: boolean;
+	hideInAdminUI?: boolean;
```
Comment (thekevinbrown) [inline src/packages/core/src/types.ts:249]:
```suggestion
```
Users can use hideInTable and hideInFilterBar together if they wish.

File: src/packages/mikroorm/src/database.ts [hunk 6]
```diff
@@ -1,4 +1,3 @@
-import 'reflect-metadata';
```
Comment (thekevinbrown) [inline src/packages/mikroorm/src/database.ts:1]: I think we might actually need this in the MikroORM package only.

File: src/packages/mikroorm/src/provider/provider.ts [hunk 7]
```diff
@@ -363,7 +363,7 @@ export class MikroBackendProvider<D extends BaseDataEntity, G extends GraphQLEnt
 		filter?: any
 	): Promise<D[]> {
 		const queryFilter = {
-			$and: [{ [relatedField]: { $in: relatedFieldIds } }, ...[filter ?? []]],
+			$and: [{ [relatedField]: { $in: relatedFieldIds } }, ...[gqlToMikro(filter) ?? []]],
```
Comment (thekevinbrown) [inline src/packages/mikroorm/src/provider/provider.ts:366]: Good catch!

File: src/packages/storage-provider/src/decorators/media-field.ts [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	BaseDataEntity,
+	Entity,
+	Field,
+	FieldOptions,
+	GraphQLEntity,
+	graphweaverMetadata,
+	FieldMetadata,
+} from '@exogee/graphweaver';
+import { S3StorageProvider } from '../storageProvider';
+import { Source } from 'graphql';
```
Comment (thekevinbrown) [inline src/packages/storage-provider/src/decorators/media-field.ts:11]: We're exporting from `graphql/types` already from `@exogee/graphweaver`. Should we also export `Source` so people don't have to have their own `graphql` dependency to get in trouble?

File: src/packages/storage-provider/src/decorators/media-field.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+	static serialize = ({ value }: { value: unknown }) => {
+		if (value === null) return null;
+		if (isMedia(value)) {
+			return JSON.stringify(
+				{
+					filename: value.filename,
+					type: value.type,
+				},
+				null,
+				2
```
Comment (thekevinbrown) [inline src/packages/storage-provider/src/decorators/media-field.ts:66]: I reckon we should save wire space here and remove the `null, 2` parameters. I know the DB will then store it however it wants, so no need to make it pretty on the way through I reckon.

File: src/packages/storage-provider/src/storageProvider.ts [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+			...(this.endpoint ? { endpoint: this.endpoint } : {}),
+		});
+
+		const fileExtension = key.split('.').pop();
+		if (!fileExtension) {
+			throw new Error('Invalid file extension');
+		}
+
+		// generate uuid from crypto module
+		const uuid = randomUUID();
```
Comment (thekevinbrown) [inline src/packages/storage-provider/src/storageProvider.ts:95]: Random thought, but if we use uuid v7, we'll get time info on the filenames too, which could be really handy: https://buildkite.com/blog/goodbye-integers-hello-uuids

**Review by thekevinbrown — APPROVED** [review]
Nothing worth blocking over, major feat!

File: src/examples/aws-cognito/package.json [hunk 11]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "@exogee/graphweaver-example-aws-cognito",
-	"version": "0.2.24",
+	"version": "1.0.0-beta-.1",
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/examples/aws-cognito/package.json:3]: Ah! I used `.` to follow semver spec:

https://semver.org/

File: src/examples/s3-storage/src/frontend/types.ts [hunk 12]
```diff
@@ -17,8 +17,8 @@ export type Scalars = {
   Boolean: { input: boolean; output: boolean; }
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/examples/s3-storage/src/frontend/types.ts:17]: done!

File: src/packages/core/src/types.ts (hunk 5 above)
Comment (steve-keep) [reply src/packages/core/src/types.ts:249]: I cant see any refs to hideInAdminUI so just removed.

File: src/packages/mikroorm/src/database.ts (hunk 6 above)
Comment (steve-keep) [reply src/packages/mikroorm/src/database.ts:1]: I think we are ok as there is a ref at the top of `src/packages/mikroorm/src/index.ts`

File: src/packages/auth/src/authentication/methods/magic-link.ts [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+		if (!link) {
+			throw new AuthenticationError('Authentication Failed: Magic Link not found');
+		}
+
+		await this.provider.updateOne(id, {
+			data: {
+				...link.data,
+				redeemedAt: new Date(),
+			},
+		});
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/auth/src/authentication/methods/magic-link.ts:172]: OK so I think there are two things:

1. Can a magic link be used twice
2. Consolidation of `redeemMagicLink` inside `verifyMagicLink`

For 1, we have this on `getMagicLink`:

```
const link = await this.provider.createOne({
			type: AuthenticationType.MagicLinkChallenge,
			userId,
			data: {
				token,
				redeemedAt: 'null',
			},
		});
```

So you cant use the same link twice once it has a date it wont come back.

Should we merge `redeemMagicLink` inside `verifyMagicLink`?

File: src/packages/storage-provider/src/decorators/media-field.ts (hunk 8 above)
Comment (steve-keep) [reply src/packages/storage-provider/src/decorators/media-field.ts:11]: done 👍

File: src/packages/storage-provider/src/storageProvider.ts (hunk 10 above)
Comment (steve-keep) [reply src/packages/storage-provider/src/storageProvider.ts:95]: Unless you are super keen I think v4 is ok is instance as we are only using it for a unique name. We could also add created timestamp to the filename if needed as it would be human readable?

File: src/examples/rest/src/backend/schema/tag.ts [hunk 14]
```diff
@@ -35,17 +30,18 @@ const acl: AccessControlList<Tag, AuthorizationContext> = {
...
 	id!: string;
 
-	@SummaryField()
 	@Field(() => String)
 	name!: string;
 
-	@RelationshipField<Task>(() => [Task], { relatedField: 'tags' })
-	tasks!: Task[];
+	@RelationshipField<Task>(() => [Task], { relatedField: 'tags', nullable: true })
+	tasks?: Task[];
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/examples/rest/src/backend/schema/tag.ts:46]: This was more of an issue on how we deal with it on the Admin UI.

The field becomes required on both sides. So when you have a many-to-many and both are required you cant create an item 😄

File: src/examples/aws-cognito/package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/examples/aws-cognito/package.json:3]: As discussed wrong version of the spec and we should remove the trailing dash so from `1.0.0-beta-.1` to `1.0.0-beta.1`

File: src/packages/storage-provider/src/storageProvider.ts (hunk 10 above)
Comment (steve-keep) [reply src/packages/storage-provider/src/storageProvider.ts:93]: As discussed as UUIDv7 is not part of crypto we will use UUIDv4 for now.

File: src/packages/auth/src/decorators/hooks/acl.ts [hunk 15]
```diff
@@ -28,7 +22,46 @@ import {
 	isPopulatedFilter,
 } from '../../auth-utils';
 
-const metadata = getMetadataStorage();
+const metadata = graphweaverMetadata;
+
+const getAccessTypeForArg = (node: unknown): AccessType => {
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/auth/src/decorators/hooks/acl.ts]: Added comments to better explain the role of this function

File: src/packages/auth/src/authentication/methods/magic-link.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/auth/src/authentication/methods/magic-link.ts]: Consolidated the extra functions into the mutation functions.

File: src/packages/storage-provider/src/storageProvider.ts (hunk 2 above)
Comment (steve-keep) [reply src/packages/storage-provider/src/storageProvider.ts:81]: Captured as part of https://exogee.atlassian.net/browse/EXOGW-355

File: src/examples/rest/src/backend/schema/tag.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/examples/rest/src/backend/schema/tag.ts:46]: https://exogee.atlassian.net/browse/EXOGW-71
